### PR TITLE
simulator UI improvements

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- text fields align to the right
- chart is larger (400px heaight)
- result summary box added
- remembers previously selected metrics to render on chart
- added ability to download chart as an image
- ability to turn off gaussian randomization
- ability for clients to exit gracefully
- added waste, max queue time, max queue time EMA
- tooltip to view data value on chart
- adds legends

## Why are we making these changes?

We want to easily compare our scaling algorithms (https://github.com/twosigma/waiter/pull/396) by rendering various metrics.

**Before**
sim-before

**After**
sim-after

### CODE REVIEW

File: waiter/resources/web/sim.html [hunk 1]
```diff
@@ -4,119 +4,200 @@
...
-				<label>scale every (secs) <input id="scale-ticks" value="5" type="number"> </label>
-				<label>simulation length (secs) <input id="total-ticks" value="3600" type="number"> </label>
+				<label>scale every (secs) <input id="scale-ticks" value="5" type="number"/></label>
+				<label>simulation length (secs) <input id="total-ticks" value="3600" type="number"/></label>
+			</div>
+        </div>
+        <div id="summary">
+			<div id="simulation-summary" class="config-box right-align">
+				<h4>Simulation Summary</h4>
+				<label>total queue time &nbsp; <input id="res-total-queue-time" value="0" type="text" readonly/></label>
```
Comment (sradack) [inline waiter/resources/web/sim.html]: We should avoid using input fields for output.

File: waiter/resources/web/sim.html [hunk 2]
```diff
@@ -4,119 +4,200 @@
...
-            $("#max-instances").val(simConfig["max-instances"]);
-        $("#scale-up-factor").val(simConfig["scale-up-factor"]);
-        $("#scale-down-factor").val(simConfig["scale-down-factor"]);
-        $("#jitter-threshold").val(simConfig["jitter-threshold"]);
-        $("#total-ticks").val(simConfig["total-ticks"]);
+        $.map($("#config input"),
+            function(el) {
+            	var metricName = $(el).attr("id");
+                if (simConfig[metricName] != null) {
+					var inputType = $(el).attr("type");
```
Comment (sradack) [inline waiter/resources/web/sim.html]: Formatting seems off here.

File: waiter/resources/web/sim.html [hunk 3]
```diff
@@ -4,119 +4,200 @@
...
         };
 
+		var currentSelectedMetrics = {};
+		$("#metrics-controls input:checked")
+			.each(function(index, el) {
+				currentSelectedMetrics[$(el).data("metric-name")] = true;
+			});
+		if ($.isEmptyObject(currentSelectedMetrics)) {
+			currentSelectedMetrics["outstanding-requests"] = true;
+            currentSelectedMetrics["total-instances"] = true;
```
Comment (sradack) [inline waiter/resources/web/sim.html]: formatting

File: waiter/resources/web/sim.html [hunk 4]
```diff
@@ -167,22 +243,46 @@ <h4>Scaler Settings</h4>
                     });
                 }
 
-                $.plot("#chart", $.map(dataSets, function(value) {
+                var totalUtilization = 0.0;
+                var totalWaste = 0.0;
+                for (var tick = 0; tick < numTicks; tick++) {
```
Comment (sradack) [inline waiter/resources/web/sim.html]: Why aren't these metrics calculated in the backend with all of the other metrics?

File: waiter/src/waiter/simulator.clj [hunk 5]
```diff
@@ -56,19 +56,28 @@
...
-  [{:strs [idle-ticks request-ticks startup-ticks]}
-   {:keys [idle-clients queued-clients idle-servers active-requests starting-servers total-queue-time
-           scale-ups scale-downs total-idle-server-time] :as current-state}
+  [{:strs [clients-exit-immediately idle-ticks randomize-times request-ticks startup-ticks]}
+   {:keys [clients-to-kill idle-clients queued-clients wait-time-max wait-time-max-ema idle-servers
+           active-requests starting-servers total-queue-time scale-ups scale-downs total-idle-server-time]
+    :as current-state}
    tick scale-amount target-instances client-change-amount]
-  {:pre [(pos? idle-ticks)
+  {:pre [(non-neg? clients-to-kill)
```
Comment (sradack) [inline waiter/src/waiter/simulator.clj]: The comment needs to be updated.

File: waiter/src/waiter/simulator.clj [hunk 6]
```diff
@@ -78,104 +87,129 @@
...
+      :total-instances total-instances
+      :total-queue-time (+ total-queue-time (count queued-clients))
+      :utilization utilization
+      :wait-time-max wait-time-max'
+      :wait-time-max-ema wait-time-max-ema'
+      :waste waste)))
 
-(let [default-initial-state {:idle-clients []
-                             :queued-clients 0
+(let [default-initial-state {:active-requests []
```
Comment (sradack) [inline waiter/src/waiter/simulator.clj]: Please keep formatting changes in separate PRs so it's straightforward to see what changed.

File: waiter/src/waiter/simulator.clj [hunk 7]
```diff
@@ -78,104 +87,129 @@
...
+                                 (min (count queued-clients) clients-to-kill)
+                                 0)
+        remaining-queued-clients (- (count queued-clients) queued-clients-to-kill)
         clients-to-kill (- clients-to-kill queued-clients-to-kill)
         interested-clients (+ remaining-queued-clients remaining-activating-clients)
         requests-created (min interested-clients available-servers)
-        requests-to-cancel (min (count (filter #(< tick %) active-requests)) clients-to-kill)
+        queued-clients' (concat queued-clients (repeat remaining-activating-clients tick))
+        wait-time-max' (->> queued-clients' (map #(- tick %)) (reduce max 0))
+        wait-time-max-ema' (+ (* 0.1 wait-time-max') (* 0.9 wait-time-max-ema))
```
Comment (sradack) [inline waiter/src/waiter/simulator.clj]: What is the purpose of the `wait-time-max-ema`?  It's not in the UI.  Do we still need it?

File: waiter/resources/web/sim.html [hunk 8]
```diff
@@ -4,119 +4,200 @@
...
 	<script src="https://cdnjs.cloudflare.com/ajax/libs/flot/0.8.3/jquery.flot.time.min.js"></script>
     <script>
 
+const numberWithCommas = (x) => {
+  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+}
+
+const showTooltip = (x, y, contents) => {
+    $('<div id="tooltip">' + contents + '</div>')
+    .css({
```
Comment (sradack) [inline waiter/resources/web/sim.html]: Please use a class and move the style details together with the rest of the style code.

File: waiter/resources/web/sim.html [hunk 9]
```diff
@@ -4,119 +4,200 @@
...
-		</label>
+		</div>
 	</body>
 
 	<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/1.12.0/jquery.min.js"></script>
 	<script src="https://cdnjs.cloudflare.com/ajax/libs/flot/0.8.3/jquery.flot.min.js"></script>
 	<script src="https://cdnjs.cloudflare.com/ajax/libs/flot/0.8.3/jquery.flot.time.min.js"></script>
     <script>
 
+const numberWithCommas = (x) => {
```
Comment (sradack) [inline waiter/resources/web/sim.html]: We're not intending to export these functions (`numberWithCommas` and `showTooltip`) so I like them inside the (function() {...})

File: waiter/resources/web/sim.html [hunk 10]
```diff
@@ -197,20 +297,28 @@ <h4>Scaler Settings</h4>
...
                 });
-                if (defaultChecked[metric])
+                if (currentSelectedMetrics[metric])
                     input.attr("checked", true);
                 $("#metrics-controls")
-                    .append($("<label/>").css("font-weight", "bold").css("color", colorMap[metric])
-                        .append(input)
-                        .append(metric));
+                    .append($("<div/>")
+                    .css("color", colorMap[metric])
```
Comment (sradack) [inline waiter/resources/web/sim.html]: Use a class and move to `<style></style>` section

File: waiter/resources/web/sim.html (hunk 4 above)
Comment (shamsimam) [reply waiter/resources/web/sim.html]: Done, moved the computations to `simulator.clj` and removed any metrics that needed computation in the UI.
